server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -99,95 +206,219 @@ public DeferringBucketCollector getDeferringCollector() {
+                liveBucketCountUnderestimate.set(owningBucketOrd, Math.toIntExact(bucketOrds.bucketsInOrd(owningBucketOrd)));
+            }
+            mergeBuckets(mergeMap, bucketOrds.size());
+            if (deferringCollector != null) {
+                deferringCollector.mergeBuckets(mergeMap);
+            }
+        }
+    }
+
     @Override
     public InternalAggregation[] buildAggregations(long[] owningBucketOrds) throws IOException {
+        correctRounding(owningBucketOrds);
+        /*
+         * Now that we have the perfect rounding rebucket everything to merge
+         * all of the buckets together that we were too lazy to merge while
+         * collecting.
+         *
+         * TODO it'd be faster if we could apply the merging on the fly as we
+         * replay the hits and build the buckets. How much faster is
+         * *interesting*. Performance tests with a couple of sub-`stats` aggs
+         * show `date_histogram` to have about the same performance as
+         * `auto_date_histogram` so there isn't really much to be gained here.
+         * But if there is a non-delaying but selectivate aggregation "above"
+         * this one then the performance gain could be substantial.
+         */
+        rebucket();
         return buildAggregationsForVariableBuckets(owningBucketOrds, bucketOrds,
                 (bucketValue, docCount, subAggregationResults) ->
                     new InternalAutoDateHistogram.Bucket(bucketValue, docCount, formatter, subAggregationResults),
-                buckets -> {
+                (owningBucketOrd, buckets) -> {
                     // the contract of the histogram aggregation is that shards must return
                     // buckets ordered by key in ascending order
                     CollectionUtil.introSort(buckets, BucketOrder.key(true).comparator());
 
                     // value source will be null for unmapped fields
                     InternalAutoDateHistogram.BucketInfo emptyBucketInfo = new InternalAutoDateHistogram.BucketInfo(roundingInfos,
-                            roundingIdx, buildEmptySubAggregations());
+                            roundingIndexFor(owningBucketOrd), buildEmptySubAggregations());
 
                     return new InternalAutoDateHistogram(name, buckets, targetBuckets, emptyBucketInfo, formatter, metadata(), 1);
                 });
     }
 
+    /**
+     * Pick the correct rounding for the specifies {@code owningBucketOrds}.
+     */
+    private void correctRounding(long[] owningBucketOrds) {
+        for (long owningBucketOrd : owningBucketOrds) {
+            byte oldRounding = roundingIndexFor(owningBucketOrd);
+            if (oldRounding >= roundingInfos.length - 1) {
+                continue;
+            }
+            byte newRounding = (byte)(oldRounding - 1);
+            long count;
+            long min = Long.MAX_VALUE;
+            long max = Long.MIN_VALUE;
+            do {
```

Comment:
Don't we have the min/max in `bounds` at this point?  Could we use that to help skip to the correct rounding, so we don't have to potentially re-build the ords several times if max-min > the threshold?

Reply:
We *do*. Let me think about that. My instinct is that this check here is largely symbolic because if the min and max would have pushed us up then we'd, well, already be on a bigger rounding anyway.